packages/server/lib/factory/IMAInternalFactory.js

```diff
@@ -0,0 +1,258 @@
+        attachment() {},
+        clearCookie() {},
+        download() {},
+        end() {},
+        format() {},
+        json() {},
+        jsonp() {},
+        links() {},
+        location() {},
+        redirect() {},
+        render() {},
+        send() {},
+        sendFile() {},
+        sendStatus() {},
+        set() {},
+        setHeader() {},
+        type() {},
+        get() {},
+        getHeader() {},
+      },
+      req: {
+        app: {},
+        baseUrl: '',
+        body: {},
+        cookies: {},
+        fresh: true,
+        hostname: '',
+        headers: {},
+        ip: '127.0.0.1',
+        ips: [],
+        method: 'GET',
+        originalUrl: '/',
+        params: {},
+        path: '',
+        protocol: 'https',
+        query: {},
+        route: {},
+        secure: true,
+        signedCookies: {},
+        stale: false,
+        subdomains: [],
+        xhr: true,
+        accepts() {},
+        acceptsCharsets() {},
+        acceptsEncodings() {},
+        acceptsLanguages() {},
+        get() {},
+        is() {},
+        range() {},
+      },
+    });
+
+    const bootConfig = createBootConfig(event);
+    const appMain = serverGlobal.get(GLOBAL.APP_MAIN);
+    event.context.app = appMain.ima.createImaApp();
+
+    emitter.emit(Event.CreateImaApp, event);
+
+    Object.assign(
+      bootConfig,
+      appMain.getInitialAppConfigFunctions(),
+      appMain.ima.getInitialPluginConfig(),
+      appMain.ima.getInitialImaConfigFunctions()
+    );
+    event.context.app.bootstrap.run(bootConfig);
+
+    return event.context.app;
+  }
+
+  function _getRouteInfo({ req, res }) {
+    let routeInfo = null;
+
+    if (!serverGlobal.has(GLOBAL.DUMMY_APP)) {
+      return routeInfo;
+    }
+
+    const dummyApp = serverGlobal.get(GLOBAL.DUMMY_APP);
+    const {
+      protocol,
+      host,
+      path: urlPath,
+      root,
+      languagePartPath,
+    } = res.locals;
+
+    dummyApp.oc.get('$Request').init(req);
+    dummyApp.oc.get('$Response').init(res);
+    dummyApp.oc.get('$Router').init({
+      $Protocol: protocol,
+      $Host: host,
+      $Path: urlPath,
+      $Root: root,
+      $LanguagePartPath: languagePartPath,
+    });
+
+    try {
+      routeInfo = dummyApp.oc.get('$Router').getCurrentRouteInfo();
+    } catch (e) {
+      console.warn('Failed to retrieve current route info', {
+        error: e,
+      });
+    }
+
+    return routeInfo;
+  }
+
+  function _addImaToResponse({ req, res }) {
+    let routeName = 'other';
+
+    let routeInfo = _getRouteInfo({ req, res });
+    if (routeInfo) {
+      routeName = routeInfo.route.getName();
+    }
+
+    res.$IMA = res.$IMA ? { ...res.$IMA, routeName } : { routeName };
+  }
+
+  function _importAppMainSync({ environment, context = {} }) {
+    let appMain = serverGlobal.has(GLOBAL.APP_MAIN)
+      ? serverGlobal.get(GLOBAL.APP_MAIN)
+      : appFactory();
+
+    if (environment.$Env === 'dev') {
+      appMain = appFactory();
```

Comment:
doesn't this call `appFactory()` twice if it's not already in the `serverGlobals`?